vignettes/ggplot2-in-packages.Rmd

```diff
@@ -0,0 +1,247 @@
+```
+
+If you use ggplot2 functions frequently, you may wish to import one or more functions from ggplot2 into your `NAMESPACE`. If you use [roxygen2](https://cran.r-project.org/package=roxygen2), you can include `#' @importFrom ggplot2 <one or more function names>` in any roxygen comment block (this will not work for datasets like `mpg`). If you do this, you will need to put ggplot2 in your `Imports` rather than your `Suggests`.
+
+```{r}
+#' @importFrom ggplot2 ggplot aes geom_bar coord_flip
+mpg_drv_summary <- function() {
+  ggplot(ggplot2::mpg) + 
+    geom_bar(aes(x = drv)) + 
+    coord_flip()
+}
+```
+
+```{r, include=FALSE}
+# make sure this function runs!
+mpg_drv_summary()
+```
+
+If you use infix operators from ggplot2 like `%+replace%` and you want to keep ggplot2 in `Suggests`, you can assign the operator within the function before it is used.
+
+```{r}
+theme_custom <- function(...) {
+  `%+replace%` <- ggplot2::`%+replace%`
+  
+  ggplot2::theme_grey(...) %+replace% 
+    ggplot2::theme(panel.background = ggplot2::element_blank())
+}
+```
+
+```{r, include=FALSE}
+# make sure this function runs!
+mpg_drv_summary() + theme_custom()
+```
+
+If you have ggplot2 in your `Imports` anyway, it is much easier to import the infix into your namespace.
+
+```{r}
+#' @importFrom ggplot2 %+replace%
+theme_custom <- function(...) {
+  ggplot2::theme_grey(...) %+replace% 
+    ggplot2::theme(panel.background = ggplot2::element_blank())
+}
+```
+
+```{r, include=FALSE}
+mpg_drv_summary() + theme_custom()
+```
+
+Even if you use many ggplot2 functions in your package, it is unwise to use ggplot2 in `Depends` or import the entire package into your `NAMESPACE`. Using ggplot2 in `Depends` will attach ggplot2 when your package is attached, which includes when your package is tested. This makes it difficult to ensure that others can use the functions in your package without attaching it (i.e., using `::`). Similarly, importing all 450 of ggplot2's exported objects into your namespace makes it difficult to separate the responsibility of your package and the responsibility of ggplot2, in addition to making it difficult for readers of your code to figure out where functions are coming from!
+
+## Using `aes()` and `vars()` in a package function
+
+To create any graphic using ggplot2 you will probably need to use `aes()` at least once. If your graphic uses facets, you might be using `vars()` to refer to columns in the plot/layer data. Both of these functions use non-standard evaluation, so if you try to use them in a function within a package they will result in a CMD check note:
+
+```{r}
+#' @importFrom ggplot2 ggplot aes geom_bar coord_flip
+mpg_drv_summary <- function() {
+  ggplot(ggplot2::mpg) + 
+    geom_bar(aes(x = drv)) + 
+    coord_flip()
+}
+```
+
+```
+N  checking R code for possible problems (2.7s)
+   mpg_drv_summary: no visible binding for global variable ‘drv’
+   Undefined global functions or variables:
+     drv
+```
+
```

Comment:
I'd add a bulleted list here that outlines the 3 options you're about to describe.

Reply:
Looks good